### ISSUE TITLE
No sound on most videos

### ISSUE THREAD

**Issue by jadesoturi** [skorokithakis/catt#211]
Hi..!

First off - awesome project!

However, I have an issue that sound is missing on 90% of the videos I play. The same video, when cast using VLC, plays just fine.

The videos are usually in .mkv H264 format.

Codec: H264 - MPEG-4 AVC (part 10) (avc1)
Codec: A/52 B Audio (aka E-AC3) (eac3)

Please advise, as this is the only solution I've found that lets me cast with working subtitles(VLC does not support casting subs - video plays, but no sub shows)

//jadesoturi

**Comment by theychx** [comment]
I do not have a surround sound setup, nor is my old tv capable of decoding EAC3, so I can't be sure what the problem is. The chromecast is capable of [audio passthrough](https://developers.google.com/cast/docs/media#audio_passthrough), which I'm guessing is why sound is working when you are using VLC.
The no sound problem when using ```catt``` is probably a limitation of the ```Default Media Receiver``` (a generic chromecast app) that we are using for casting non-YouTube content (I assume the VLC people have made their own chromecast app).

For now, I'm afraid the only solution would be to transcode audio to plain stereo aac.

**Comment by theychx** [comment]
..or maybe playback with VLC is working, because it is transcoding audio on-the-fly.

**Comment by jadesoturi** [comment]
Hi and thank for the reply. 

I don't have a surround setup for this either. The chromecast is connected to a hdmi hub that has ARC and then sends the sound out through a regular minijack to some creative speakers. The HDMI switch also has SPDIF/PASSTHROUGH/5.1 and I CAN in theory connect it to a receiver through a TOS-link to get surround. 

AFAIK VLC does not transcode on the fly, as it hardly uses any resource when playing back the video(1.6% on a old DualCore), so I will try to reencode the video, but what format would you recommend?

**Comment by theychx** [comment]
> but what format would you recommend?

Keep the video stream and re-encode audio to aac (or mp3 if you prefer). With ```ffmpeg``` you would do:
```
ffmpeg -i movie.mkv -c:v copy -c:a aac reencoded_movie.mkv
```

**Comment by skorokithakis** [comment]
Are we sure it's not just because the Chromecast doesn't support E-AC-3 decoding? It seems like it only supports passthrough.

**Comment by jadesoturi** [comment]
Well, I find it odd that VLC works then. It does not transcode on-the-fly. I did re-encode the video, but I think it has trouble with 5.1 sound, cause now its garbled even when playing locally. I will mess abit more with the bitrate and see if I can downmix it properly. Also, since it supports passthrough, it should work when I use passthrough on the HDMI switch, no?

**Comment by skorokithakis** [comment]
VLC works for casting, or just for playing?

**Comment by jadesoturi** [comment]
I can cast the original .mkv with eac3 in VLC.

**Comment by skorokithakis** [comment]
Hmm, that's very odd... I wonder if it actually *is* doing transcoding, since it only needs to transcode at 1x realtime it wouldn't use that much CPU, but I don't know if it would use as little as 1.6%. It must be, since the Chromecast doesn't look like it supports the codec... I wonder if there's a way to see which codec it's actually using right now.

**Comment by jadesoturi** [comment]
It starts out at around 4% according to gotop, and drops to around 2% with a few minutes. This is when i use the GUI and select my Chromecast from the Playback - Renderer list.

If i use cvlc to run it instead, it uses around 5-10%:

 vlc videofile.mkv --sout "#chromecast" --sout-chromecast-ip=192.168.0.105 --demux-filter=demux_chromecast

Its odd that it uses more CPU when not running the GUI, but maybe it uses different settings there?

**Comment by theychx** [comment]
> Are we sure it's not just because the Chromecast doesn't support E-AC-3 decoding? It seems like it only supports passthrough

Indeed. What I was trying to say before was, that it seems that any given cc-app (including the ```Default Media Receiver```) needs to [implement](https://developers.google.com/cast/docs/caf_receiver_overview) a mechanism that handles passthrough, and that the ```Default Media Receiver``` might not implement this. Or, it could be that it actually does, but that we can't utilize it from ```pychromecast```.

**Comment by jadesoturi** [comment]
VLC identifies as Default Media Receiver when casting, so I guess it is implemented since I get sound on VLC. As for pychromecast, that might be. Know far to little about it. I will try some more today, see if I can find a video that works with catt and see if it's eac3. However, pretty sure most videos I got are h264 with eac3.

**Comment by gilemk** [comment]
Just wondering if any progress on this issue has been made? I just found out about CATT and I am excited about it. But eventually ran into this sound issue.

**Comment by skorokithakis** [comment]
AFAIK we arrived at this being a Chromecast issue, no? My Chromecast has been playing videos fine, but I don't think any of my videos use E-AC3 for sound.

**Comment by gilemk** [comment]
Well, when I play the video via VLC it playes. But when I do it via CATT I have no sound. So I do not see how this is a chromecast issue. So if you know anything I might be missing, let me know.

**Comment by skorokithakis** [comment]
I think VLC transcodes the video.

**Comment by gilemk** [comment]
You are probably right. I do still want to use CATT for casting to my chromecast though. So what about enabling the pass-through with CAF as previously mentioned in this issue?

**Comment by gilemk** [comment]
From my own testing, seems AC-3 works fine, but for E-AC-3 I get no sound. Looking at the docs it says:

_The Cast Application Framework (CAF) supports AC-3 passthrough. If you're not using CAF, you can check the codec support prior to creating the source buffer by using the CastReceiverContext.canDisplayType() method and passing mp4a.a5 for AC-3 or mp4a.a6 for E-AC-3._

So maybe all is needed is to check whether a codec is E-AC-3 and pass it as mp4a.a6?

**Comment by theychx** [comment]
@gilemk 
> So what about enabling the pass-through with CAF as previously mentioned in this issue?

> So maybe all is needed is to check whether a codec is E-AC-3 and pass it as mp4a.a6?

The code that is utilizing CAF runs on the chromecast, not the box running `catt`. So we would have to make our own cc-app. Doable, but certainly not trivial.

**Comment by zbvk** [comment]
having the same issue, chromecast seems to support only a handful of audio codecs and converting them into AAC does sound like a hastle, any workarounds?

**Comment by alexandervlpl** [comment]
I have this issue with **A52 Audio (aka AC3) (a52)**. VLC casts these files without problems, but for `catt` I need to convert to aac.

**Comment by alexandervlpl** [comment]
If anyone else is struggling with this, I've found that even `aac` does not work reliably for me. Sometimes `catt` just crashes immediately after starting a cast and nothing is played.

What does work is `libmp3lame`, and encodes much faster too! Audiophiles may not be happy but I really doubt you'll hear the difference on any home system.

```
ffmpeg -i movie.mkv -c:v copy -c:a libmp3lame reencoded_movie.mkv
```

**Comment by tekknolagi** [comment]
libmp3lame did it for me when nothing else did!
